### ISSUE TITLE
[Async TP] all-gather-matmuls not fusing properly when rowwise scales are used

### ISSUE THREAD

**Issue by danielvegamyhre** [pytorch/pytorch#149990]
## Summary
I recently implemented support for rowwise scales for fusing matmul-reduce-scatter patterns in async TP: https://github.com/pytorch/pytorch/pull/149247 

When testing the performance of various configurations to ensure stability of the changes, I found that while float8 rowwise training with async TP had correct numerical accuracy, the performance was non-optimal (see benchmarks below).

After looking at the traces, I found the matmul-reduce-scatters were being fused properly, so my change was working as intended - however, the all-gather-matmul patterns were NOT being fused properly. This seems to (at least in part) explain the poor performance for async TP with rowwise scales.

However, looking at the benchmarks below we also see vanilla TP perf with rowwise scales is unexpectedly low. I will create a separate issue for this, though.

### Performance Benchmarks

Llama3 70b training runs on 128 H100s with full AC, using FSDP=16, TP=8
- **bf16 (vanilla TP):** 598 TPS, peak memory 71.51 GB
- **bf16 (async TP):** TPS  673, peak memory 71.08 (+12.54% TPS vs vanilla TP)
- **float8 tensorwise (vanilla TP):** 820 TPS, peak memory  55.26 GB
- **float8 tensorwise (async TP):** 950 TPS, peak memory 55.91 GB (+15.85% TPS vs vanilla TP)
- **float8 rowwise (vanilla TP):** TPS: 540 TPS, peak memory 71.46 GB
- **float8 rowwise (async TP):** 560 TPS, peak memory 70.65 GB (+3.7% TPS vs vanilla TP but still unexpectedly lower than bf16)

As you can see, float8 rowwise is working but performance needs to be improved further.

cc @vkuzo @lessw2020 

### Versions

pytorch @ HEAD

**Comment by danielvegamyhre** [comment]
I'm working on this but cc-ing distributed folks to ensure they have visibility on this as well @tianyu-l @kwen2501

**Comment by danielvegamyhre** [comment]
I looked into the forward pass and found all gather matmuls are fusing for bf16 and tensorwise properly because the "find_consumer_matmuls" function works properly, it finds 3 scaled_mm ops using the view/reshape op at the end of the matched all-gather pattern. However, for rowwise, the output of "cat" goes into some float8 conversion ops before the view/reshape -> scaled_mm, rather than the view/reshape -> scaled_mm:

![Image](https://github.com/user-attachments/assets/ad18dc2c-75ee-4d18-89fa-bdac6bd7d464)

**Comment by danielvegamyhre** [comment]
cc @drisspg as well if you have any thoughts on the potential solution outlined here: https://github.com/pytorch/pytorch/issues/149990#issuecomment-2763860155

**Comment by drisspg** [comment]
One clarification question, if the PerTensor was communicated in bf16 and then converted would this also not cause the pattern to match?

**Comment by danielvegamyhre** [comment]
> One clarification question, if the PerTensor was communicated in bf16 and then converted would this also not cause the pattern to match?

Yes, if the conversion to float8 happens after the all-gather, the pattern will not match.

**Comment by vkuzo** [comment]
> However, one option we could look into is supporting rowwise scaling for these. My understanding is that there's no benefit of doing all-gather in float8 w/ rowwise scaling, since the overhead of (float8 conversion + M number of scales per tensor instead of 1, so many more bytes sent over the network) >= just sending the data in bf16.

I think an estimate of eng time and expected benefit would help decide if this is worth doing! Thoughts on what those might be?

**Comment by danielvegamyhre** [comment]
update: I prototyped a solution and discussed ideas with @fegin but there are some issues:

1) Supporting fp8 all gather in TP for rowwise scales seems like it will be pretty ugly, implementation wise, due to needing 2 different "input_hp" and "grad_output" tensors (row-major with scales along logical rows, and col-major with scales along logical columns). The prototype still is not fully working but in theory it should be possible.

2) The "input" tensor is 3d, and rowwise quant primitives only support 2d shapes (axiswise dim in {0,-1}). We can reshape to 2D before the scaling => all-gather (in Float8RowwiseParallel/Float8ColwiseParallel) , but then the metadata of the original tensor shape is lost, and in the Float8Linear.forward() we don't know how to reshape it back to the 3d+ shape.

3) What if the user shards along an axis that we are computing scales along? We need to validate against this somewhere.

**Comment by danielvegamyhre** [comment]
update: I have a prototype fp8 rowwise all-gather working for the llama3 debug model with 1 layer, but there is still a bug for larger models.

**Comment by danielvegamyhre** [comment]
update: i managed to get fp8 rowwise all-gather working for llama3 8b. perf is flat compared to using bf16 all-gather, which is expected, since we are transferring 2 fp8 `input` tensors over the network, instead of 1 bf16 `input` tensor. however, this should allow us to integrate with async TP.

**Comment by danielvegamyhre** [comment]
i've found that when using Float8RowwiseParallel and Float8ColumnwiseParallel, there are actually no all-gathers happening in float8, it must be some kind of other collective in *.redistribute()? because we have an assertion that scales are tensorwise in all_gather, wait_tensor, cat, split op overrides, and that assertion is not triggered.

Furthermore, I found looking at the graph that fp8 quant ops are still occuring after the all gathers, so implementing Float8RowwiseParallel and Float8ColwiseParallel is not sufficient, we must implement PrepareFloat8ModuleInput as well, and add support for rowwise scales in primitive op overrides: all_gather, wait_tensor, cat, split.

**Comment by danielvegamyhre** [comment]
another update: I've found that when Attention.forward() has an optional 2nd input (col-major) and passes that through to the projections (`self.wq(*x)` etc, ) which are Float8Linears - which now accept this optional 2nd arg -  for some reason, a pre-forward module hook is converting the 2nd arg from a `DTensor(Float8Tensor)` => None? This is causing a problem.

**Comment by danielvegamyhre** [comment]
Latest issue: I have a PrepareModuleInput subclass for float8 rowwise, which is creating has an "input_layout" of Shard(dim=1) and "desired_layout" of Replicate(). So I expect an all-gather along dim=1. However, it all-gathers along dim=0. I set a breakpoint and confirmed this in the op override:


```python
@implements(
    [
        c10d_functional.all_gather_into_tensor.default,
        _c10d_functional.all_gather_into_tensor.default,
    ]
)
def allgather_fp8(aten_op, args, kwargs=None):
    """
    override funcol with FP8 handling
    """
    #_assert_tensorwise_scale(aten_op, args[0]._scale)
    torch.distributed.breakpoint()
    fp8_input = args[0] # (8, 1024, 256)
    assert isinstance(fp8_input, Float8Tensor), (
        f"expecting a Float8Tensor for allgather but found {type(fp8_input)}"
    )

    fp8_data = fp8_input._data
    fp8_data = fp8_data.contiguous()

    # fp8_out becomes (16, 1024, 256) <- gathered along dim 0 even though gather_dim is 1?
    # input_layout = Shard(dim=1), desired_layout=Replicate()
    fp8_out = aten_op(fp8_data, *args[1:], **kwargs) 
    return Float8Tensor(
        fp8_out,
        fp8_input._scale,
        fp8_input._orig_dtype,
        fp8_input._linear_mm_config,
        fp8_input._gemm_input_role,
    )

```

**Comment by danielvegamyhre** [comment]
Update: i have the full e2e PoC working! however, the loss is spiky / not identical to the bf16 run, so looking into that.

**Comment by danielvegamyhre** [comment]
Update: figured out the issue, input-col-major should have 0 grads since it is only created for rowwise all gather and not meaningful for the gradient flow, so i did some tricks to pass in a 0 filled DTensor, and now the loss curve looks good.

**Comment by danielvegamyhre** [comment]
Another update: I started looking into why async TP is crashing w/ fp8 rowwise all gather, and found it's due to a similar issue to https://github.com/pytorch/pytorch/pull/149247 where the "A" tensor used is from before a reshape, and the "A_scale" node is from after a reshape, so there's a mismatch.

**Comment by danielvegamyhre** [comment]
another update: strangely, the loss curve looked good for the first 100 steps, yet I think the scales were actually wrong - I was repeating the scale for shard0 across both shards (TP degree=2) and somehow it was working. After fixing this and using the local shards for each tensor, it works for eager mode but with compile I get NaNs....

**Comment by danielvegamyhre** [comment]
> another update: strangely, the loss curve looked good for the first 100 steps, yet I think the scales were actually wrong - I was repeating the scale for shard0 across both shards (TP degree=2) and somehow it was working. After fixing this and using the local shards for each tensor, it works for eager mode but with compile I get NaNs....

Update (forgot to post this last week): I figured out the reason for this, basically it seems to me there is a fundamental incompatibility between fp8 with rowwise scales in torchao and async TP. I stepped through the various collective ops with a debugger and discovered an issue with how all-gather works under the hood. This is the relevant [code](https://github.com/pytorch/pytorch/blob/6f6fac6a416b64221031d32e6c1e4df145df45f5/torch/distributed/_functional_collectives.py#L210-L215).

To illustrate the problem I created this diagram:

![Image](https://github.com/user-attachments/assets/4515a7ec-7d3d-4d76-a4ed-738c1919998a)

Basically no matter what the sharding/gather dim is defined as (e.g. `Shard(dim=1) => Replicate()`), it does the following:
1. all-gather along dim=0 (presumably due to how NCCL works, or perhaps it is just more efficient for collectives to write to row major layout - I know async TP also forces collectives to be along dim 0 as well, so I assume these collective primitives are more efficient this way)
2. if gather dim != 0, then:
    - chunk along dim=0
    - cat along gather dim
  
This is problematic for the case where we need to convert the input activation tensor to float8 in row-major format with scales along logical rows (for `output = input @ weight.t()`) AND convert to float8 in col-major format with scales along logical columns (for `grad_weight = grad_output_t @ input`). 

As you can see in the diagram, we cannot all-gather fp8 input tensor shards when the scale dim is the same as the all-gather dim (i.e., the case of the input shard in col-major format with scales along logical columns). The torchao Float8Tensor absraction cannot represent this. 

The reason we were getting NaNs is because I was doing concatenating the float8 shards to set Float8Tensor._data, but setting the Float8Tensor._scale as the maxes of the scales, which is incorrect, since the original scales have already been applied to the fp8 data, so when we dequantize using the wrong scales we can get NaNs.

IMO we need to develop a solution to this that will also be compatible with MX formats + async TP, as well as future proof against future quantization strategies and dtypes.

**Possible solutions**
1. Extend Float8Tensor or make a new tensor subclass with a more flexible way to represent "scale => subtensor associated with that scale." This is more like a workaround imo and would add additional complexity, debugability issues, etc.
2. Add new async TP pattern matching to include the dynamic quant ops, and directly implement them in the custom ops which replace the target subgraph. We can't import torchao in PT core and have that be a dependency, so this would replicating the work we've done in torchao and maintaining parity between the two (not good).
3. **Preferred option**: Have inductor post grad pass and potentially custom symmetric memory ops for implementations of async TP for low precision training (float8, MX) live in torchao. This is my preferred option because:      
     - We can include the dynamic quant ops in the pattern matching without issue, since we can directly use the quantization code in torchao in the custom ops replacing the target subgraph.
     - We won't be forced to try to make float8/MX/etc subgraphs fit the same pattern matching as bf16 without breaking anything, which is very limiting and/or not possible in some cases (like this one). 
         - For example, the only reason fp8 tensorwise training is compatible with async TP is because we can do all-gather in fp8 and thus the inductor pattern matches, but if we constrain async TP to only being compatible with dynamic quantization strategies that can be all-gathered in low-precision, that is limiting both composability and peak achievable performance.
     - Still a consistent story for the user of using torchao for "all things quantization and sparsity"
     - There is already precedent for doing this in torchao: https://github.com/pytorch/ao/pull/2140

**Comment by danielvegamyhre** [comment]
@vkuzo @drisspg any thoughts on option 3 in the above comment (https://github.com/pytorch/pytorch/issues/149990#issuecomment-2861706351)? I'd like to align internally with the team first before proposing it to PTD team soon for H2 planning

**Comment by andoorve** [comment]
Hey @vkuzo @danielvegamyhre any update on this issue?

**Comment by andoorve** [comment]
Thanks!

> FYI we are deprioritizing fp8 rowwise training in favor of mxfp8, and likely nvfp4 soon (for B200), as well as fp8 blockwise (for H100) due to industry/usage trends.

Isn't it the case that all of these formats will run into the exact same issues described here? I.e. we still require AG -> Quantize -> GEMM to happen the same way for all of these formats.
